# Applying Advanced Transformations 
Student Name: Nathalie Castro

In [29]:
import json
import numpy as np
import pandas as pd

### Data Loading

In [30]:
pd.set_option('display.max_columns', None)

In [31]:
powers_url = "https://docs.google.com/spreadsheets/d/e/2PACX-1vQ2cMaGI74T_DqeCWDojRIyUCQqNZxsu_L3f42JJKV-_f873im-CBttJU8qn_Kan22qa71uCXfDWdMI/pub?output=csv"
# Load the data
powers_df = pd.read_csv(powers_url)
# Verify it loaded
powers_df.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [32]:
info_url = "https://docs.google.com/spreadsheets/d/e/2PACX-1vRgXxC_AvDsxnrZSMIuKnPMvrSrL6ZnoLogMNU-ZD8uOB88pbsINNtkii1PlMTgWPSyrqm5JJnU6pCr/pub?output=csv"
# Load the data
info_df = pd.read_csv(info_url)
# Verify it loaded
info_df.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

### Data Cleaning

In [33]:
# Check for missing values
info_df.isna().sum().sum()

0

In [34]:
# Check for missing values
powers_df.isna().sum().sum()

0

In [35]:
# Check for duplicates
info_df.duplicated().sum()

0

In [36]:
# Check for duplicates
powers_df.duplicated().sum()

0

In [37]:
# Split the "Hero|Publisher" column using split and save the 
info_df[["Hero", "Publisher"]] = info_df["Hero|Publisher"].str.split("|", expand=True)
# Drop the original column
info_df.drop("Hero|Publisher", axis=1, inplace=True)
# Verify it worked
info_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [38]:
# Separate Height From Weight
a = info_df.loc[0, "Measurements"]
type(json.loads(a.replace("'", '"')))

dict

In [39]:
info_df["Measurements"] = info_df["Measurements"].str.replace("'", '"')
# Apply JSON loads to the whole column to make the 
# string into a dictionary
info_df["Measurements"] = info_df["Measurements"].apply(json.loads)
# Verify it worked
info_df["Measurements"].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [40]:
# Create a DataFrame with just the height and weights
heights_weights = info_df["Measurements"].apply(pd.Series)
heights_weights.head()

Height    Weight
0  203.0 cm  441.0 kg
1  191.0 cm   65.0 kg
2  185.0 cm   90.0 kg
3  203.0 cm  441.0 kg
4  193.0 cm  122.0 kg

In [41]:
# Rename the columns of the height weight dataframe
heights_weights.rename(columns = {"Height": "Height (cm)", "Weight": "Weight (kg)"}, inplace=True)
heights_weights.head()

Height (cm) Weight (kg)
0    203.0 cm    441.0 kg
1    191.0 cm     65.0 kg
2    185.0 cm     90.0 kg
3    203.0 cm    441.0 kg
4    193.0 cm    122.0 kg

In [42]:
# Remove units from the measurements strings then cast each as a float
heights_weights["Height (cm)"] = heights_weights["Height (cm)"].str.split(" ", expand=True)[0].astype(float)
heights_weights["Weight (kg)"] = heights_weights["Weight (kg)"].str.split(" ", expand=True)[0].astype(float)
# Verify floats
heights_weights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Height (cm)  463 non-null    float64
 1   Weight (kg)  463 non-null    float64
dtypes: float64(2)
memory usage: 7.4 KB


In [43]:
# Add the heights_weights dataframe to the original
info_df = pd.concat([info_df, heights_weights], axis=1)
info_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  Height (cm)  Weight (kg)  
0      Marvel Comics        203.0        441.0  
1  Dark Horse Comics        191.0         65.0  
2          DC Comics        185.0         90.0  
3      Marvel Comics        203.0        441.0  
4      Marvel Comics        193.0        122.0

In [44]:
# Remove the original "Measurements" column
info_df.drop(columns="Measurements", inplace=True)
info_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height (cm)  Weight (kg)  
0         A-Bomb      Marvel Comics        203.0        441.0  
1     Abe Sapien  Dark Horse Comics        191.0         65.0  
2       Abin Sur          DC Comics        185.0         90.0  
3    Abomination      Marvel Comics        203.0        441.0  
4  Absorbing Man      Marvel Comics        193.0        122.0

In [45]:
# Make the columns into a list
columns = list(info_df.columns)
columns

['Gender',
 'Race',
 'Alignment',
 'Hair color',
 'Eye color',
 'Skin color',
 'Hero',
 'Publisher',
 'Height (cm)',
 'Weight (kg)']

In [46]:
# Rearrange columns
columns = [columns[columns.index("Hero")], \
           columns[columns.index("Publisher")], \
           columns[columns.index("Alignment")],\
           columns[columns.index("Gender")], \
           columns[columns.index("Race")], \
           columns[columns.index("Height (cm)")], \
           columns[columns.index("Weight (kg)")], \
           columns[columns.index("Skin color")], \
           columns[columns.index("Eye color")], \
           columns[columns.index("Hair color")]]
columns

['Hero',
 'Publisher',
 'Alignment',
 'Gender',
 'Race',
 'Height (cm)',
 'Weight (kg)',
 'Skin color',
 'Eye color',
 'Hair color']

In [47]:
# Apply the new rearrangement to the old info_df
info_df = info_df[columns]
info_df.head()

Hero          Publisher Alignment Gender               Race  \
0         A-Bomb      Marvel Comics      good   Male              Human   
1     Abe Sapien  Dark Horse Comics      good   Male      Icthyo Sapien   
2       Abin Sur          DC Comics      good   Male            Ungaran   
3    Abomination      Marvel Comics       bad   Male  Human / Radiation   
4  Absorbing Man      Marvel Comics       bad   Male              Human   

   Height (cm)  Weight (kg) Skin color Eye color Hair color  
0        203.0        441.0    Unknown    yellow    No Hair  
1        191.0         65.0       blue      blue    No Hair  
2        185.0         90.0        red      blue    No Hair  
3        203.0        441.0    Unknown     green    No Hair  
4        193.0        122.0    Unknown      blue    No Hair

### OHE 

In [48]:
# Split the "Powers" string by commas into a list
powers_df["Powers Split"] = powers_df["Powers"].str.split(",")
powers_df.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        Powers Split  
0    [Agility, Super Strength, Stamina, Super Speed]  
1  [Accelerated Healing, Durability, Longevity, S...  
2  [Agility, Accelerated Healing, Cold Resistance...  
3                               [Lantern Power Ring]  
4  [Accelerated Healing, Intelligence, Super Stre...

In [49]:
# Create a new DataFrame of exploded powers
powers_exploded = powers_df.explode("Powers Split")
# Find the unique powers in the exploded column
unique_powers = np.sort(powers_exploded["Powers Split"].unique())
unique_powers

array(['Accelerated Healing', 'Adaptation', 'Agility',
       'Animal Attributes', 'Animal Control', 'Animal Oriented Powers',
       'Animation', 'Anti-Gravity', 'Astral Projection', 'Astral Travel',
       'Audio Control', 'Banish', 'Biokinesis', 'Camouflage',
       'Changing Armor', 'Clairvoyance', 'Cloaking', 'Cold Resistance',
       'Cryokinesis', 'Danger Sense', 'Darkforce Manipulation',
       'Death Touch', 'Density Control', 'Dexterity',
       'Dimensional Awareness', 'Dimensional Travel', 'Duplication',
       'Durability', 'Echolocation', 'Elasticity', 'Electrical Transport',
       'Electrokinesis', 'Element Control',
       'Elemental Transmogrification', 'Empathy', 'Energy Absorption',
       'Energy Armor', 'Energy Beams', 'Energy Blasts',
       'Energy Constructs', 'Energy Manipulation', 'Energy Resistance',
       'Enhanced Hearing', 'Enhanced Memory', 'Enhanced Senses',
       'Enhanced Sight', 'Enhanced Smell', 'Enhanced Touch',
       'Fire Control', 'Fire Resis

In [50]:
# Create each power column and add true/false values 
for power in unique_powers:
    power_df = pd.DataFrame({power: powers_df["Powers"].str.contains(power)})
    powers_df = pd.concat([powers_df, power_df], axis=1)
powers_df.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        Powers Split  Accelerated Healing  \
0    [Agility, Super Strength, Stamina, Super Speed]                False   
1  [Accelerated Healing, Durability, Longevity, S...                 True   
2  [Agility, Accelerated Healing, Cold Resistance...                 True   
3                               [Lantern Power Ring]                False   
4  [Accelerated Healing, Intelligence, Super Stre...                 True   

   Adaptation  Agility  Animal Attributes  Animal Control  \
0       False     True              False           False   
1       False    False              False           False   
2       False     True              False           False   
3       False    False              False           False   
4       False    False              False           False   

   Animal Oriented Powers  Animation  Anti-Gravity  Astral Projection  \
0                   False      False         False              False   
1                   False      False         False              False   
2                   False      False         False              False   
3                   False      False         False              False   
4                   False       True         False              False   

   Astral Travel  Audio Control  Banish  Biokinesis  Camouflage  \
0          False          False   False       False       False   
1          False          False   False       False        True   
2          False          False   False       False       False   
3          False          False   False       False       False   
4          False          False   False       False       False   

   Changing Armor  Clairvoyance  Cloaking  Cold Resistance  Cryokinesis  \
0           False         False     False            False        False   
1           False         False     False            False        False   
2           False         False     False             True        False   
3           False         False     False            False        False   
4           False         False     False            False        False   

   Danger Sense  Darkforce Manipulation  Death Touch  Density Control  \
0         False                   False        False            False   
1         False                   False        False            False   
2         False                   False        False            False   
3         False                   False        False            False   
4         False                   False        False            False   

   Dexterity  Dimensional Awareness  Dimensional Travel  Duplication  \
0      False                  False               False        False   
1      False                  False               False        False   
2      False                  False               False        False   
3      False                  False               False        False   
4      False                  False               False        False   

   Durability  Echolocation  Elasticity  Electrical Transport  Electrokinesis  \
0       False         False       False                 False           False   
1        True         False       False                 False           False   
2        True         False       False                 False           False   
3       False         False       False                 False           False   
4       False         False       False                 False           False   

   Element Control  Elemental Transmogrification  Empathy  Energy Absorption  \
0            False         

In [51]:
# Drop the now unnecessary Power and Power Split columns
powers_df.drop(columns=["Powers", "Powers Split"], inplace=True)
powers_df.head()

hero_names  Accelerated Healing  Adaptation  Agility  Animal Attributes  \
0      3-D Man                False       False     True              False   
1       A-Bomb                 True       False    False              False   
2   Abe Sapien                 True       False     True              False   
3     Abin Sur                False       False    False              False   
4  Abomination                 True       False    False              False   

   Animal Control  Animal Oriented Powers  Animation  Anti-Gravity  \
0           False                   False      False         False   
1           False                   False      False         False   
2           False                   False      False         False   
3           False                   False      False         False   
4           False                   False       True         False   

   Astral Projection  Astral Travel  Audio Control  Banish  Biokinesis  \
0              False          False          False   False       False   
1              False          False          False   False       False   
2              False          False          False   False       False   
3              False          False          False   False       False   
4              False          False          False   False       False   

   Camouflage  Changing Armor  Clairvoyance  Cloaking  Cold Resistance  \
0       False           False         False     False            False   
1        True           False         False     False            False   
2       False           False         False     False             True   
3       False           False         False     False            False   
4       False           False         False     False            False   

   Cryokinesis  Danger Sense  Darkforce Manipulation  Death Touch  \
0        False         False                   False        False   
1        False         False                   False        False   
2        False         False                   False        False   
3        False         False                   False        False   
4        False         False                   False        False   

   Density Control  Dexterity  Dimensional Awareness  Dimensional Travel  \
0            False      False                  False               False   
1            False      False                  False               False   
2            False      False                  False               False   
3            False      False                  False               False   
4            False      False                  False               False   

   Duplication  Durability  Echolocation  Elasticity  Electrical Transport  \
0        False       False         False       False                 False   
1        False        True         False       False                 False   
2        False        True         False       False                 False   
3        False       False         False       False                 False   
4        False       False         False       False                 False   

   Electrokinesis  Element Control  Elemental Transmogrification  Empathy  \
0           False            False                         False    False   
1           False            False                         False    False   
2           False            False                         False    False   
3           False            False                         False    False   
4           False            False                         False    False   

   Energy Absorption  Energy Armor  Energy Beams  Energy Blasts  \
0              False         False         False          False   
1              False         False         False          False   
2              False         False         False          False   
3              False         False         False          False   
4              False         False         False          False   

   Energy C

In [52]:
# Merge the tables for values where "Hero" = "hero_names" and then drop the "hero_names" column
final_df = pd.merge(info_df, powers_df, left_on="Hero", right_on="hero_names").drop(columns="hero_names")
final_df

Hero          Publisher Alignment  Gender               Race  \
0             A-Bomb      Marvel Comics      good    Male              Human   
1         Abe Sapien  Dark Horse Comics      good    Male      Icthyo Sapien   
2           Abin Sur          DC Comics      good    Male            Ungaran   
3        Abomination      Marvel Comics       bad    Male  Human / Radiation   
4      Absorbing Man      Marvel Comics       bad    Male              Human   
..               ...                ...       ...     ...                ...   
458     Yellowjacket      Marvel Comics      good    Male              Human   
459  Yellowjacket II      Marvel Comics      good  Female              Human   
460             Yoda       George Lucas      good    Male     Yoda's species   
461          Zatanna          DC Comics      good  Female              Human   
462             Zoom          DC Comics       bad    Male            Unknown   

     Height (cm)  Weight (kg) Skin color Eye color        Hair color  \
0          203.0        441.0    Unknown    yellow           No Hair   
1          191.0         65.0       blue      blue           No Hair   
2          185.0         90.0        red      blue           No Hair   
3          203.0        441.0    Unknown     green           No Hair   
4          193.0        122.0    Unknown      blue           No Hair   
..           ...          ...        ...       ...               ...   
458        183.0         83.0    Unknown      blue             Blond   
459        165.0         52.0    Unknown      blue  Strawberry Blond   
460         66.0         17.0      green     brown             White   
461        170.0         57.0    Unknown      blue             Black   
462        185.0         81.0    Unknown       red             Brown   

     Accelerated Healing  Adaptation  Agility  Animal Attributes  \
0                   True       False    False              False   
1                   True       False     True              False   
2                  False       False    False              False   
3                   True       False    False              False   
4                  False       False    False              False   
..                   ...         ...      ...                ...   
458                False       False    False              False   
459                False       False    False              False   
460                False       False     True              False   
461                False       False    False              False   
462                False       False    False              False   

     Animal Control  Animal Oriented Powers  Animation  Anti-Gravity  \
0             False                   False      False         False   
1             False                   False      False         False   
2             False                   False      False         False   
3             False                   False       True         False   
4             False                   False      False         False   
..              ...                     ...        ...           ...   
458           False                    True      False         False   
459           False                   False      False         False   
460           False                   False      False         False   
461           False                   False      False         False   
462           False                   False      False         False   

     Astral Projection  Astral Travel  Audio Control  Banish  Biokinesis  \
0                False          False          False   False       False   
1                False          False          False   False       False   
2                False          False          False   False       False   
3                False          False          False   False       False   
4                False          False          False   False       False   
..                 ...            ...            ...

### Compare the average weight of super powers who have Super Speed to those who do not.

In [53]:
# Group the dataframe by super speed and find the mean of the weight
final_df.groupby('Super Speed')[['Weight (kg)']].mean()

Weight (kg)
Super Speed             
False         101.773585
True          129.404040

Average weight for someone with super speed is 129.40 kg. 

### What is the average height of heroes for each publisher?

In [54]:
final_df.groupby("Publisher")[["Height (cm)"]].mean().sort_values(by="Height (cm)")

Height (cm)
Publisher                     
George Lucas        159.600000
Shueisha            171.500000
Dark Horse Comics   176.909091
Unknown             178.000000
Team Epic TV        180.750000
Star Trek           181.500000
DC Comics           181.923913
Marvel Comics       191.546128
Image Comics        211.000000

The average height for heroes in each publisher is between 160 kg to 211 k